File: hrl/advait_sandbox/arm_navigation_tutorials/src/get_state_validity.cpp


#include <ros/ros.h>
#include <planning_environment_msgs/GetStateValidity.h>

int main(int argc, char **argv){
  ros::init (argc, argv, "get_state_validity_test");
  ros::NodeHandle rh;

  ros::service::waitForService("environment_server_right_arm/get_state_validity");
  ros::ServiceClient check_state_validity_client_ = rh.serviceClient<planning_environment_msgs::GetStateValidity>("environment_server_right_arm/get_state_validity");

  planning_environment_msgs::GetStateValidity::Request req;
  planning_environment_msgs::GetStateValidity::Response res;

  req.robot_state.joint_state.name.push_back("r_shoulder_pan_joint");
  req.robot_state.joint_state.name.push_back("r_shoulder_lift_joint");
  req.robot_state.joint_state.name.push_back("r_upper_arm_roll_joint");
  req.robot_state.joint_state.name.push_back("r_elbow_flex_joint");
  req.robot_state.joint_state.name.push_back("r_forearm_roll_joint");
  req.robot_state.joint_state.name.push_back("r_wrist_flex_joint");
  req.robot_state.joint_state.name.push_back("r_wrist_roll_joint");
  req.robot_state.joint_state.position.resize(7,0.0);

  //these set whatever non-zero joint angle values are desired
  //req.robot_state.joint_state.position[0] = 0.0;
  //req.robot_state.joint_state.position[0] = -.55;
  req.robot_state.joint_state.position[0] = 0.40;
  req.robot_state.joint_state.position[3] = -0.40;
 
  req.robot_state.joint_state.header.stamp = ros::Time::now();
  req.check_collisions = true;
  if(check_state_validity_client_.call(req,res))
  {
    if(res.error_code.val == res.error_code.SUCCESS)
      ROS_INFO("Requested state is not in collision");
    else
      ROS_INFO("Requested state is in collision. Error code: %d",res.error_code.val);
  }
  else
  {
    ROS_ERROR("Service call to check state validity failed %s",check_state_validity_client_.getService().c_str());
    return false;
  }
  ros::shutdown();
}

File: hrl/advait_sandbox/arm_navigation_tutorials/src/arm_navigation_tutorials/get_state_validity.py

import roslib; roslib.load_manifest('arm_navigation_tutorials')
import sys
import rospy

from planning_environment_msgs.srv import GetStateValidity
from planning_environment_msgs.srv import GetStateValidityRequest


if __name__ == '__main__':
    rospy.init_node('get_state_validity_python')

    srv_nm = 'environment_server_right_arm/get_state_validity'
    rospy.wait_for_service(srv_nm)
    get_state_validity = rospy.ServiceProxy(srv_nm, GetStateValidity)

    req = GetStateValidityRequest()
    req.robot_state.joint_state.name = ['r_shoulder_pan_joint',
                                        'r_shoulder_lift_joint',
                                        'r_upper_arm_roll_joint',
                                        'r_elbow_flex_joint',
                                        'r_forearm_roll_joint',
                                        'r_wrist_flex_joint',
                                        'r_wrist_roll_joint']
    req.robot_state.joint_state.position = [0.] * 7
    req.robot_state.joint_state.position[0] = 0.4
    req.robot_state.joint_state.position[3] = -0.4

    req.robot_state.joint_state.header.stamp = rospy.Time.now()
    req.check_collisions = True

    res = get_state_validity.call(req)
    
    if res.error_code.val == res.error_code.SUCCESS:
        rospy.loginfo('Requested state is not in collision')
    else:
        rospy.loginfo('Requested state is in collision. Error code: %d'%(res.error_code.val))



